File: src/624.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int maxDistance(vector<vector<int>>& arrays) {
        int retVal = 0;

        int arraysSize = arrays.size();

        int arraysColSize = arrays[0].size();
        int minValue = arrays[0][0];
        int maxValue = arrays[0][arraysColSize - 1];
        for (int i = 1; i < arraysSize; ++i) {
            arraysColSize = arrays[i].size();
            int diff1 = abs(arrays[i][arraysColSize - 1] - minValue);
            int diff2 = abs(maxValue - arrays[i][0]);
            retVal = max(retVal, max(diff1, diff2));

            minValue = min(minValue, arrays[i][0]);
            maxValue = max(maxValue, arrays[i][arraysColSize - 1]);
        }

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<vector<int>> arrays;
    };
    vector<subject> testData{{{{1, 2, 3}, {4, 5}, {1, 2, 3}}}, {{{1}, {1}}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: arrays = [[1,2,3],[4,5],[1,2,3]]
     *  Output: 4
     *
     *  Input: arrays = [[1],[1]]
     *  Output: 0
     */

    Solution cSolution;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: arrays = [";
        for (long unsigned int j = 0; j < testData[i].arrays.size(); ++j) {
            cout << (j == 0 ? "" : ",") << "[";
            for (long unsigned int k = 0; k < testData[i].arrays[j].size(); ++k) {
                cout << (k == 0 ? "" : ",") << testData[i].arrays[j][k];
            }
            cout << "]";
        }
        cout << "]\n";

        answer = cSolution.maxDistance(testData[i].arrays);
        cout << "Output: " << answer << "\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/624.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def maxDistance(self, arrays: List[List[int]]) -> int:
        retVal = 0

        arraysSize = len(arrays)

        minValue = arrays[0][0]
        maxValue = arrays[0][-1]
        for i in range(1, arraysSize):
            diff1 = abs(arrays[i][-1] - minValue)
            diff2 = abs(maxValue - arrays[i][0])
            retVal = max(retVal, diff1, diff2)

            minValue = min(minValue, arrays[i][0])
            maxValue = max(maxValue, arrays[i][-1])

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for arrays in [[[1, 2, 3], [4, 5], [1, 2, 3]], [[1], [1]]]:
            # /* Example
            #  *  Input: arrays = [[1,2,3],[4,5],[1,2,3]]
            #  *  Output: 4
            #  *
            #  *  Input: arrays = [[1],[1]]
            #  *  Output: 0
            #  */
            logging.info("Input: arrays = %s", arrays)

            retVal = pSolution.maxDistance(arrays)
            logging.info("Output: %d", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
